src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp

```diff
@@ -1275,6 +1275,14 @@ HeapWord* ShenandoahHeap::allocate_new_gclab(size_t min_size,
 HeapWord* ShenandoahHeap::allocate_new_plab(size_t min_size,
                                             size_t word_size,
                                             size_t* actual_size) {
+  size_t unalignment = min_size % CardTable::card_size_in_words();
+  if (unalignment != 0) {
+    min_size = min_size - unalignment + CardTable::card_size_in_words();
+  }
+  unalignment = word_size % CardTable::card_size_in_words();
+  if (unalignment != 0) {
+    word_size = word_size - unalignment + CardTable::card_size_in_words();
+  }
```

Comment:
Probably not a big deal since this is only used when refilling a PLAB, which is an infrequent operation, but `mod` is an expensive operation, in general, and best to avoid in our code except in assertion checks (or even there given recent experiences with debug tests timing out). Since card size is a power of 2, may be we could use addition and masking instead. Something like defining the following inline in the CardTable class and using it everywhere where card alignment granularity is sought. There may even be a macro or method defined for this already perhaps:

```
(FOO + CardSize - 1) & ~((1 << LogCardSize) - 1)
```

One could even store the mask to avoid the arithmetic to produce the mask although it's pretty cheap.

This may turn out to be less expensive than mod, test, and branch, but as I said probably not a big deal here. We should make sure we don't overuse mods in our allocation paths much.

Reply:
Thanks for this suggestion.  I've modified the code.